app/src/main/java/com/example/stretchy/features/executetraining/sound/SoundPlayer.kt

```diff
@@ -0,0 +1,204 @@
+    private val players: MutableList<MediaPlayer> = mutableListOf()
+    override val soundFlow: MutableSharedFlow<SoundFlow> = MutableSharedFlow(replay = 0)
+    val soundList = mutableListOf<SoundType>()
+
+    init {
+        textToSpeech = TextToSpeech(context) { status ->
+            if (status == TextToSpeech.SUCCESS) {
+                val languageResult = textToSpeech.setLanguage(locale)
+                if (languageResult == TextToSpeech.LANG_MISSING_DATA || languageResult == TextToSpeech.LANG_NOT_SUPPORTED) {
+                    Log.e(TAG, "Language not supported")
+                } else {
+                    val voice = Voice(
+                        "en-us-x-tpc-local",
+                        Locale("en", "US"),
+                        1000,
+                        0,
+                        true,
+                        emptySet()
+                    )
+                    textToSpeech.voice = voice
+                    textToSpeech.setPitch(pitch)
+                    textToSpeech.setSpeechRate(speechSpeed)
+
+
+                    GlobalScope.launch {
+                        isInitializedFlow.emit(true)
+                    }
+                    CoroutineScope(Dispatchers.IO).launch {
+                        soundFlow.collect {
+                            Log.e("soundflow",it.toString())
+                        }
+                    }
+                }
+            }
+        }
+
+    }
+
+    fun queueSound(sound: SoundType) {
+        when (sound) {
+            is SoundType.Sound -> {
+                soundList.add(SoundType.Sound(sound.soundTrack))
+                if (soundList.size == 1) {
+                    playSound(sound.soundTrack)
+                }
+            }
+            is SoundType.Speech -> {
+                if(soundList.getOrNull(0) is SoundType.Speech){
+                    soundList.removeAt(0)
+                    textToSpeech.stop()
+                }else if(soundList.getOrNull(0) == SoundType.Sound(SoundTrack.EXERCISE_ENDING)){
+                    soundList.removeAt(0)
+                    stopSound()
+                }
+                soundList.add(SoundType.Speech(sound.text))
+                if (soundList.size == 1) {
+                    CoroutineScope(Dispatchers.IO).launch {
+                        sound.text?.let { say(it) }
+                    }
+                }
+            }
+        }
+
+    }
+
+    fun stopSound() {
+        for (player in players) {
+            if (player.isPlaying) {
+                player.stop()
+                player.reset()
+                player.release()
+            }
+        }
+        players.clear()
+    }
+
+
+    private suspend fun say(text: String) {
+        isInitializedFlow
+            .filter { it }
+            .first()
+
+        textToSpeech.speak(
+            text,
+            TextToSpeech.QUEUE_FLUSH, null, "UniqueID"
+        )
+        textToSpeech.setOnUtteranceProgressListener(object :
```

Comment:
this should not be set here, we've already discussed that. If you set the listener here you basically override existing listeners. I guess this is the source of bugs